lib/fluent/event_router.rb

```diff
@@ -136,23 +136,85 @@ class Pipeline
       def initialize
         @filters = []
         @output = nil
+        @optimizer = FilterOptimizer.new
       end
 
       def add_filter(filter)
         @filters << filter
+        @optimizer.filters = @filters
       end
 
       def set_output(output)
         @output = output
       end
 
       def emit_events(tag, es)
-        processed = es
-        @filters.each { |filter|
-          processed = filter.filter_stream(tag, processed)
-        }
+        processed = @optimizer.filter_stream(tag, es)
         @output.emit_events(tag, processed)
       end
+
+      class FilterOptimizer
+        def initialize(filters = [])
+          @filters = filters
+        end
+
+        def filters=(filters)
+          @filters = filters
+          reset_optimization
+        end
+
+        def filter_stream(tag, es)
+          if optimizable?
+            optimized_filter_stream(tag, es)
+          else
+            $log.info "Can't optimized filters, because some plugins implement `filter_stream`"
```

Comment:
Log messages should be for users, not only for plugin developers.
So it's better to show "filtering works with worse performance, 'xxx_filter' users #filter_stream method" or so.
Users will get known about performance, and can complaint on that plugin about implementation.